src/src/Helpers/Modbus_link.cpp

```diff
@@ -0,0 +1,210 @@
+
+  if (ModbusLINK_struct::_easySerial == nullptr) {
+    return false;
+  }
+
+  // Set RS485 mode if requested using selected pin for RTS
+  bool rs485Mode = ModbusLINK_struct::_easySerial->setRS485Mode(dere_pin, collision_detect);
+
+  ModbusLINK_struct::_easySerial->begin(baudrate);
+
+  if (loglevelActiveFor(LOG_LEVEL_DEBUG)) {
+    String log = F("P183: Init serial: RX pin ");
+    log += serial_rx;
+    log += F(", TX pin ");
+    log += serial_tx;
+    log += F(", RS485 mode selected on pin ");
+    log += dere_pin;
+    log += F(", baudrate ");
+    log += baudrate;
+    log += F(", collision detection ");
+    log += collision_detect ? F("enabled") : F("disabled");
+    log += F(", RS485mode enabled: ");
+    log += rs485Mode ? F("yes") : F("no");
+    addLogMove(LOG_LEVEL_DEBUG, log);
+  }
+
+  return true;
+}
+
+// Return the initialization status of the link
+bool ModbusLINK_struct::isInitialized() const {
+  return _easySerial != nullptr;
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Queue a Modbus request. The request is appended to the request and assigned a unique identifier.
+// The client can use this identifier to retrieve the response later.
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+uint16_t ModbusLINK_struct::queueRequest(struct ModbusDEVICE_struct *device,
+                                         uint8_t                    *sendframe,
+                                         uint16_t                    sendframe_length,
+                                         uint16_t                    rcvframe_length,
+                                         uint16_t                    timeout) {
+  struct Modbus_RequestQueueElement *req = new Modbus_RequestQueueElement(_queueID, 0);
+
+  req->_id     = ++(_queueID);                 // Assign a unique ID to the request
+  req->_device = device;                       // The Modbus device making the request
+
+  for (int i = 0; i < sendframe_length; ++i) { // Make a copy of the sendframe to persist it in teh queue element
+    req->_sendframe[i] = sendframe[i];
+  }
+  req->_sendframe_length = sendframe_length;
+  req->_rcvframe_length  = rcvframe_length;
+  req->_timeout          = timeout;
+  req->_state            = 0;    // Initial state
+
+  _requestQueue.push_back(*req); // Append the request to the queue
+  return req->_id;
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Check if there is a response available for the given request ID and retrieve it if available
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+bool ModbusLINK_struct::getResponse(uint16_t id, uint8_t *rcvframe, uint16_t rcvframe_length) {
```

Comment:
This function has a very deep `if...else return` structure.
When you invert the if checks you can immediately return.
This makes the function much easier to read and also way less stack usage.

Reply:
Refactored, still working on a proper handling of the the message queue. This function will be deleted as all processing of the queue is now in processCommand. That might require refactoring also...